include/sys/btree.h

```diff
@@ -115,7 +120,41 @@ typedef struct btree {
 	uint64_t		bt_num_nodes;
 	zfs_btree_hdr_t		*bt_root;
 	zfs_btree_leaf_t	*bt_bulk; // non-null if bulk loading
-} zfs_btree_t;
+};
+
+/*
+ * Implementation of Shar's algorithm designed to accelerate binary search by
+ * eliminating impossible to predict branches. For optimality, this should be
+ * used to generate the search function in the same file as the comparator and
+ * the comparator must be marked `__attribute__((always_inline) inline` so that
+ * the compiler can inline the search function.
+ *
+ * Consumers must opt into using this function by setting ->bt_find_in_buf
```

Comment:
This comment is out of date and should be removed in my next push. Otherwise, this should be the final version.